add StepRange support for CartesianIndices

johnnychen94 opened the pull request:
This PR intended to only add StepRange support for CartesianIndices, but more things are added here as I was working on it.

Three major changes are involved here:

## StepRange CartesianIndices

The biggest change here is to make it possible to pass a `step` to CartesianIndices so that more flexible loop behavior can be done

```julia
R = CartesianIndices(A)
Δ = CartesianIndex(ntuple(_->2, ndims(A)))

for I in first(R):Δ:last(R)
    ...
end
```

Another new thing introduced by this PR is that `step(::CartesianIndices)` now returns a `CartesianIndex`, this enables us to treat CartesianIndices as a multidimensional range type.

## Reduce CartesianIndices overhead

By reworking the iterating part, the overhead issue #9080 is accidentally fixed.

```julia
using Test
using BenchmarkTools

function sumcart_iter(A)
    s = zero(eltype(A))
    for I in CartesianIndices(A)
        @inbounds s += A[I]
    end
    s
end
function sumcart_manual(A::AbstractMatrix)
    s = zero(eltype(A))
    ax1, ax2 = axes(A)
    for j in ax2, i in ax1
        @inbounds s += A[i, j]
    end
    s
end


```

```julia
julia> A = rand(4, 4);

julia> @btime sumcart_iter($A)
  6.757 ns (0 allocations: 0 bytes) # this PR
  15.315 ns (0 allocations: 0 bytes) # master

julia> @btime sumcart_manual($A) # baseline
  8.013 ns (0 allocations: 0 bytes)

julia> A = rand(256,20);

julia> @btime sumcart_iter($A);
  4.543 μs (0 allocations: 0 bytes) # this PR
  5.017 μs (0 allocations: 0 bytes) # master

julia> @btime sumcart_manual($A);
  4.544 μs (0 allocations: 0 bytes) # baseline
```

So :tada: looping over `CartesianIndices` is marginally faster than a manual loop.

There's still some overhead for StepRange:

```julia
function sumcart_iter(A)
    s = zero(eltype(A))
    R = CartesianIndices(A)
    Δ = CartesianIndex(ntuple(_->2, ndims(R)))
    for I in first(R):Δ:last(R)
        @inbounds s += A[I]
    end
    s
end

function sumcart_manual(A::AbstractMatrix)
    s = zero(eltype(A))
    ax1, ax2 = axes(A)
    for j in first(ax2):2:last(ax2), i in first(ax1):2:last(ax1)
        @inbounds s += A[i, j]
    end
    s
end
```

```julia
julia> A = rand(8, 8);

julia> @btime sumcart_iter($A)
  36.418 ns (0 allocations: 0 bytes)

julia> @btime sumcart_manual($A)
  32.492 ns (0 allocations: 0 bytes)
```

but that's more difficult to deal with because (IIUC) `step` can be either positive or negative. I'm not very familiar with this kind of low-level optimization...

## tests and bug fixes

A lot of tests are spread in other files, I tried to collect most of the test patterns into `test/cartesian.jl` so it is easier to first check if `include("test/cartesian.jl")` works before triggering the whole heavy `make test` workflow.

A bug is found and fixed (see also #37928) :

```julia
# master
julia> CartesianIndices((2, 1:2))
1×2 CartesianIndices{2, Tuple{UnitRange{Int64}, UnitRange{Int64}}}:
 CartesianIndex(2, 1)  CartesianIndex(2, 2)

# this PR
julia> CartesianIndices((2, 1:2))
2×2 CartesianIndices{2, Tuple{Base.OneTo{Int64}, UnitRange{Int64}}}:
 CartesianIndex(1, 1)  CartesianIndex(1, 2)
 CartesianIndex(2, 1)  CartesianIndex(2, 2)

# master/this PR
julia> CartesianIndices((2, 2))
2×2 CartesianIndices{2, Tuple{Base.OneTo{Int64}, Base.OneTo{Int64}}}:
 CartesianIndex(1, 1)  CartesianIndex(1, 2)
 CartesianIndex(2, 1)  CartesianIndex(2, 2)
```


New tests are added to test common operations and properties. Old tests are still kept so as to not break anything unexpectedly.

---

There is still something I want to tackle, but I'm a bit running out of my spare time to keep working on this, which I'll leave them as future work:

- StepRange support to LinearIndices (https://github.com/JuliaLang/julia/pull/37829#discussion_r499183657)
- tweak the StepRange CartesianIndices overhead

cc: @timholy @mbauman 

closes #37577 
closes #9080

johnnychen94 committed:
[WIP] add StepRange support for CartesianIndices

johnnychen94 commented on base/multidimensional.jl:
```diff
@@ -395,11 +422,8 @@ module IteratorsMD
+    @inline in(i::CartesianIndex{N}, r::CartesianIndices{N}) where {N} = all(map(in, i.I, r.indices))
 
     simd_outer_range(iter::CartesianIndices{0}) = iter
     function simd_outer_range(iter::CartesianIndices)
```
I don't know if it's needed to modify this simd part:

https://github.com/JuliaLang/julia/blob/8a9666ae2278c61e021ed21864feb77d04a53dc7/base/multidimensional.jl#L412-L415

timholy replied:
Probably we're missing a test. "Inner" iteration in a `@simd` goes from `0:n-1`, you'll need to compute the correct index from that.

johnnychen94 replied:
Could you give me some test examples that I can work on? I don't quite know what should be the correct behavior...

johnnychen94 replied:
Changed in 7e1b9aa but I'm not sure if I get things done correctly.

johnnychen94 replied:
6efb112 fixes the simd performance by propogating inbounds check.

johnnychen94 commented on base/multidimensional.jl:
```diff
@@ -351,37 +376,39 @@ module IteratorsMD
-    @inline function inc(state, start, stop)
-        _, I = __inc(state, start, stop)
+    # TODO: optimize this; it adds up about 5ns overhead
+    @inline function inc(state, start, step, stop)
```
A better strategy to remove the overhead that I have in mind is to dispatch `inc`/`dec` on `StepRange` and `AbstractUnitRange`.

johnnychen94 replied:
Surprisingly, a0e64c3 also reduces the overhead that #35074 tries to solve.

```julia
function sumcart_iter(A)
    s = zero(eltype(A))
    for I in CartesianIndices(A)
        @inbounds s += A[I]
    end
    s
end
function sumcart_manual(A::AbstractMatrix)
    s = zero(eltype(A))
    ax1, ax2 = axes(A)
    for j in ax2, i in ax1
        @inbounds s += A[i, j]
    end
    s
end
A = rand(4, 4)
using Test
@test sumcart_iter(A) == sumcart_manual(A)
using BenchmarkTools
@btime sumcart_iter($A)
@btime sumcart_manual($A)
```

```julia
julia> @btime sumcart_iter($A)
  7.316 ns (0 allocations: 0 bytes) # this PR
  14.839 ns (0 allocations: 0 bytes) # master
8.549672595595762
julia> @btime sumcart_manual($A)
  8.107 ns (0 allocations: 0 bytes)
8.549672595595762
```

and for the step range version:

```julia
function sumcart_iter(A)
    s = zero(eltype(A))
    R = CartesianIndices(A)
    Δ = CartesianIndex(ntuple(_->2, ndims(R)))
    for I in first(R):Δ:last(R)
        @inbounds s += A[I]
    end
    s
end

function sumcart_manual(A::AbstractMatrix)
    s = zero(eltype(A))
    ax1, ax2 = axes(A)
    for j in first(ax2):2:last(ax2), i in first(ax1):2:last(ax1)
        @inbounds s += A[i, j]
    end
    s
end

A = rand(8, 8)
using Test
@test sumcart_iter(A) == sumcart_manual(A)
using BenchmarkTools
@btime sumcart_iter($A)
@btime sumcart_manual($A)
```

```julia
julia> @btime sumcart_iter($A)
  15.162 ns (0 allocations: 0 bytes)
9.210002105899234

julia> @btime sumcart_manual($A)
  27.261 ns (0 allocations: 0 bytes)
9.210002105899234
```

johnnychen94 commented on base/multidimensional.jl:
```diff
@@ -395,11 +422,8 @@ module IteratorsMD
-    _in(b, ::Tuple{}, ::Tuple{}, ::Tuple{}) = b
-    @inline _in(b, i, start, stop) = _in(b & (start[1] <= i[1] <= stop[1]), tail(i), tail(start), tail(stop))
+    @inline in(i::CartesianIndex, r::CartesianIndices) = false
+    @inline in(i::CartesianIndex{N}, r::CartesianIndices{N}) where {N} = all(map(in, i.I, r.indices))
```
this gives a better performance:

```julia
julia> R = CartesianIndices((1:3, 1:4))
3×4 CartesianIndices{2, Tuple{UnitRange{Int64}, UnitRange{Int64}}}:
 CartesianIndex(1, 1)  CartesianIndex(1, 2)  …  CartesianIndex(1, 4)
 CartesianIndex(2, 1)  CartesianIndex(2, 2)     CartesianIndex(2, 4)
 CartesianIndex(3, 1)  CartesianIndex(3, 2)     CartesianIndex(3, 4)

julia> @btime $CartesianIndex(1) in $R
  0.046 ns (0 allocations: 0 bytes) # this PR
  1.429 ns (0 allocations: 0 bytes) # master

julia> @btime $CartesianIndex(1, 2) in $R
  0.046 ns (0 allocations: 0 bytes) # this PR/master
```

timholy replied:
The result is spurious; to take 1/20th of ns you'd need a 20GHz CPU (so that 1 clock cycle = 0.05ns), but AFAIK such CPUs don't exist. Try using the `setup=(i = CartesianIndex(1))` notation in BenchmarkTools and see if that helps.

johnnychen94 replied:
LOL it seems that is still the same:

```julia
julia> @btime i in $R setup=(i=CartesianIndex(1, 2))
  0.046 ns (0 allocations: 0 bytes)
true

julia> @btime i in $R setup=(i=CartesianIndex(1))
  0.044 ns (0 allocations: 0 bytes)
false
```

Should I revert the `in(i::CartesianIndex, r::CartesianIndices) = false` method?

timholy replied:
That one might be there to catch cases where someone asks whether a `CartesianIndex{3}` is an element in a `CartesianIndices{2}` or something.

johnnychen94 commented on base/multidimensional.jl:
```diff
@@ -281,17 +301,26 @@ module IteratorsMD
     convert(::Type{Tuple{}}, R::CartesianIndices{0}) = ()
-    convert(::Type{NTuple{N,AbstractUnitRange{Int}}}, R::CartesianIndices{N}) where {N} =
-        R.indices
+    for RT in (OrdinalRange{Int, Int}, StepRange{Int, Int}, AbstractUnitRange{Int})
+        @eval convert(::Type{NTuple{N,$RT}}, R::CartesianIndices{N}) where {N} =
+            R.indices
+    end
```
I'm not sure if this done in the correct way... the nested multi-level dispatch is still a bit confusing to me.

johnnychen94 commented on base/multidimensional.jl:
```diff
@@ -248,29 +261,36 @@ module IteratorsMD
-    CartesianIndices(inds::NTuple{N,Union{<:Integer,AbstractUnitRange{<:Integer}}}) where {N} =
-        CartesianIndices(map(i->first(i):last(i), inds))
+    CartesianIndices(inds::NTuple{N,Union{<:Integer,OrdinalRange{<:Integer}}}) where {N} =
+        CartesianIndices(map(_range2ind, inds))
```
I feel this is more of a bug fix:

```julia
# master
julia> CartesianIndices((2, 1:2))
1×2 CartesianIndices{2, Tuple{UnitRange{Int64}, UnitRange{Int64}}}:
 CartesianIndex(2, 1)  CartesianIndex(2, 2)

# this PR
julia> CartesianIndices((2, 1:2))
2×2 CartesianIndices{2, Tuple{Base.OneTo{Int64}, UnitRange{Int64}}}:
 CartesianIndex(1, 1)  CartesianIndex(1, 2)
 CartesianIndex(2, 1)  CartesianIndex(2, 2)
```

timholy reviewed (commented):
I think this is a good change. But the biggest worry about this is that a lot of code written for CartesianIndices has formerly assumed that the step size would always be 1, and that will no longer be true. Sort of like the transition to indexing starting at something other than 1, this could trigger a bigger shakeup across the ecosystem. It's probably the right thing to do, but be aware that this change may have a "tail" that could last for years. (It's probably a lot less disruptive than the 1-indexing change, though.)

One thing we can probably do to ease the transition is define
```julia
const UnitCartesianIndices{N,R<:NTuple{N,AbstractUnitRange{Int}}} = CartesianIndices{N,R}
```
and encourage people to use it when they want to enforce a step size of 1. (Not sure whether to export this, though, probably not.)

johnnychen94 committed:
optimize iterate

johnnychen94 committed:
apply suggestions

johnnychen94 committed:
add step to simd_index

johnnychen94 reviewed (commented):
Where should the new test goes, `test/cartesian.jl`?

johnnychen94 committed:
fix convert

johnnychen94 committed:
add constructors to OrdinalRange

johnnychen94 committed:
fix simd performance

johnnychen94 commented on base/multidimensional.jl:
```diff
@@ -248,29 +262,36 @@ module IteratorsMD
         indices::R
     end
 
     CartesianIndices(::Tuple{}) = CartesianIndices{0,typeof(())}(())
```
This seems a little bit inconsistant:

```julia
julia> CartesianIndices(())
0-dimensional CartesianIndices{0, Tuple{}}:
CartesianIndex()

julia> CartesianIndices(rand(0))
0-element CartesianIndices{1, Tuple{Base.OneTo{Int64}}}
```

johnnychen94 replied:
I think this inconsistency comes from the fact that:

```julia
julia> axes(())
(Base.OneTo(0),)

julia> length(())
0

julia> size(())
ERROR: MethodError: no method matching size(::Tuple{})
```

In any case, it should be fixed in other PRs.

johnnychen94 committed:
add tests

Missing tests for AbstractUnitRange is also added

johnnychen94 committed:
add reverse for CartesianIndices

johnnychen94 commented on base/multidimensional.jl:
```diff
@@ -448,36 +469,38 @@ module IteratorsMD
     iterate(iter::Reverse{<:CartesianIndices{0}}, state=false) = state ? nothing : (CartesianIndex(), true)
 
-    Base.LinearIndices(inds::CartesianIndices{N,R}) where {N,R} = LinearIndices{N,R}(inds.indices)
+    Base.LinearIndices(inds::CartesianIndices{N,R}) where {N,R} = LinearIndices(axes(inds))
```
To actually support StepRange, `LinearIndices` should be expanded in the same manner, too.

This change isn't correct:

```julia
julia> A = rand(1:10, 7, 7);

julia> R = CartesianIndex(1, 1):CartesianIndex(2, 3):CartesianIndex(7, 7);

julia> A[R] == A[LinearIndices(R)]
false
```

johnnychen94 committed:
various fixes

* correct overflow checks
* correct broadcasting

johnnychen94 committed:
add a temp patch to make sure test passes

johnnychen94 commented on base/multidimensional.jl:
```diff
@@ -343,45 +368,75 @@ module IteratorsMD
+    @propagate_inbounds function __inc(state::Tuple{Int}, indices::Tuple{<:OrdinalRange})
+        rng = indices[1]
+        I = state[1] + step(rng)
+        valid = __isvalid_inc_range(I, rng) && __has_inc_overflow(state[1], rng)
```
A confusing name:

```suggestion
       valid = __isvalid_inc_range(I, rng) && __no_inc_overflow(state[1], rng)
```

timholy replied:
Yeah, those two do seem to mean pretty different things.

timholy commented on base/multidimensional.jl:
```diff
@@ -248,29 +262,36 @@ module IteratorsMD
     CartesianIndices(index::CartesianIndex) = CartesianIndices(index.I)
-    CartesianIndices(sz::NTuple{N,<:Integer}) where {N} = CartesianIndices(map(Base.OneTo, sz))
-    CartesianIndices(inds::NTuple{N,Union{<:Integer,AbstractUnitRange{<:Integer}}}) where {N} =
-        CartesianIndices(map(i->first(i):last(i), inds))
```
I'm not sure why this "normalized" to UnitRanges, that seems inconsistent (I like your version better), but keep an eye out for weird bugs. I seem to remember seeing some difficulities when `axes(A)` returned a heterogeneous tuple.

johnnychen94 replied:
This change is reverted (just to minimize the changes of this PR)

```diff
     _convert2ind(sz::Integer) = Base.OneTo(sz)
-    _convert2ind(sz::OrdinalRange) = sz
+    _convert2ind(sz::AbstractUnitRange) = first(sz):last(sz)
+    _convert2ind(sz::OrdinalRange) = first(sz):step(sz):last(sz)
```

timholy commented on base/multidimensional.jl:
```diff
@@ -343,45 +368,75 @@ module IteratorsMD
+    @propagate_inbounds function __inc(state::Tuple{Int}, indices::Tuple{<:OrdinalRange})
+        rng = indices[1]
+        I = state[1] + step(rng)
+        valid = __isvalid_inc_range(I, rng) && __has_inc_overflow(state[1], rng)
```
Wondering if we should use use the iteration of the underlying ranges? I'm no longer remembering why it was implemented this way.

johnnychen94 replied:
Yes, this is a better idea.

There are some uncaught StepRange overflow behavior #37876 and breaks some of the existing tests, I'll mark them as broken.

johnnychen94 replied:
> I'm no longer remembering why it was implemented this way.

Here's my guess. `iterate` protocol returns `Union{Nothing, Tuple}`, the current implementation eliminates this type instability by passing an explicit `valid` flag, which makes the whole `__inc` thing is type stable.

The version I tried locally is

```julia
    iterate(iter::CartesianIndices, state) = _maybe_cartesian(__iterate(iter.indices, state.I))
    _maybe_cartesian(::Nothing) = nothing
    _maybe_cartesian(I::Tuple) = CartesianIndex(I), CartesianIndex(I)

    __iterate(indices::Tuple{<:OrdinalRange}, state::Tuple{Int}) = _maybe_int(iterate(indices[1], state[1]))
    _maybe_int(::Nothing) = nothing
    _maybe_int(I::Tuple) = (I[1], )

    @propagate_inbounds function __iterate(indices, state)
        rng = indices[1]
        first_idx = _iterate_range(rng, state[1])
        if !isnothing(first_idx)
            # the first column still has elements
            return (first_idx, tail(state)...)
        end
        # otherwise, recursively iterate on other columns
        first_idx = first(rng)
        tailing_idxs = __iterate(tail(indices), tail(state))
        if !isnothing(tailing_idxs)
            return (first_idx, tailing_idxs...)
        end
        return nothing
    end

    @propagate_inbounds _iterate_range(rng::OrdinalRange, state) = __iterate_result(iterate(rng, state))
    __iterate_result(::Nothing) = nothing
    __iterate_result(t::Tuple) = first(t)
```

as you can see, three one-liners are defined to extract and bypass `nothing`, but this still shakes the entire types in `__iterate`.

As a comparison, 

```julia
R = CartesianIndices((0:1, 0:2))
i = CartesianIndex(1, 0)
```

```
julia> @btime iterate(R, i)
158.875 ns (4 allocations: 112 bytes) # my local version
22.060 ns (1 allocation: 48 bytes) # the current version
```

timholy commented on base/multidimensional.jl:
```diff
@@ -440,44 +492,78 @@ module IteratorsMD
+            indices = map(ind->convert(AbstractUnitRange, ind), indices)
+            LinearIndices{N, typeof(indices)}(indices)
+        else
+            throw(ArgumentError("LinearIndices for $typeof(inds) with non-1 step size is not supported."))
```
Actually, thinking about this more, maybe we can handle this? After all, the definition of `LinearIndices` is
```julia
struct LinearIndices{N,R<:NTuple{N,AbstractUnitRange{Int}}} <: AbstractArray{Int,N}
    indices::R
end
```
and so some of the same tricks you're using here could be used there too. You might have to reset the integer at the end of each "column," perhaps by maintaining a hybrid linear/cartesian iterator:
```julia
julia> inds = LinearIndices((1:3, 1:4))
3×4 LinearIndices{2, Tuple{UnitRange{Int64}, UnitRange{Int64}}}:
 1  4  7  10
 2  5  8  11
 3  6  9  12

julia> inds[1:2:3,:]   # can't just increase by a constant
2×4 Matrix{Int64}:
 1  4  7  10
 3  6  9  12
```
Only, of course, when it's a StepRange.

Keep in mind that converting Cartesian to Linear is pretty fast (multiply and add); it's the opposite transition, Linear to Cartesian, that requires `div` and hence is dirt-slow. (That's why Cartesian is the fallback.)

timholy commented on test/cartesian.jl:
```diff
@@ -19,25 +230,89 @@ ex = Base.Cartesian.exprresolve(:(if 5 > 4; :x; else :y; end))
+        @test iterate(I, i) === nothing
+    end
+
+    @testset "decremental" begin
```
This word is sufficiently unfamiliar (at least to me) that perhaps saying `"decremental ranges"` or something would help.

timholy commented on test/cartesian.jl:
```diff
@@ -7,6 +7,217 @@ ex = Base.Cartesian.exprresolve(:(if 5 > 4; :x; else :y; end))
+        (CartesianIndices((1:2:5, )), CartesianIndex(2, ), CartesianIndex(4, )),
+        (1:2:5, 2, 4),
+    ]
+        # consistant to ranges
```
```suggestion
        # consistent with ranges
```

timholy reviewed (commented):
A very impressive PR! Solving #9080 alone is amazing, but doing so while cleaning up implementations, adding more flexibility, testing more thoroughly is a tour de force. 

Subject to fixing the test errors this seems basically ready from what I can tell, pending how you want to resolve the LinearIndices situation. To be clear, I think that could be a separate PR, if you want to go with the errorring version here first and get it out in the wild.

johnnychen94 committed:
apply suggestions

johnnychen94 committed:
fix collect

johnnychen94 commented:
Does `step(R::CartesianIndices) = CartesianIndex(step.(R.indices))` be crazy to you? With this PR done, I feel it might be pretty useful to think `CartesianIndices` as a multidimensional range now.

johnnychen94 committed:
loose overflow checks on StepRange CartesianIndices

To keep consistent with the StepRange behavior

johnnychen94 committed:
some cleanup

johnnychen94 committed:
add step method

johnnychen94 commented on base/multidimensional.jl:
```diff
@@ -440,44 +480,69 @@ module IteratorsMD
+            indices = map(ind->convert(AbstractUnitRange, ind), indices)
+            LinearIndices{N, typeof(indices)}(indices)
+        else
+            throw(ArgumentError("LinearIndices for $typeof(inds) with non-1 step size is not yet supported."))
```
```suggestion
            throw(ArgumentError("LinearIndices for $(typeof(inds)) with non-1 step size is not yet supported."))
```

johnnychen94 commented on base/multidimensional.jl:
```diff
@@ -344,47 +369,65 @@ module IteratorsMD
-    @inline function __inc(state::Tuple{Int}, start::Tuple{Int}, stop::Tuple{Int})
-        valid = state[1] < stop[1]
-        return valid, (state[1]+1,)
+    # Unlike ordinary ranges, CartesianIndices requires the iteration continue on the next column
+    # when there is no valid item in the current column.
+    # `iterate` returns `Union{Nothing, Tuple}`, we explicitly pass a `valid` flag to eliminate
+    # the type instability inside the core `__inc` logic, and this gives better runtime performance.
```
Based on the discussion, I add some comments so that future readers get some clue of what this part of the code is doing.

johnnychen94 commented on base/multidimensional.jl:
```diff
@@ -440,44 +480,69 @@ module IteratorsMD
+
+    # This is needed because converting to LinearIndices is only available when steps are all 1
+    # NOTE: this is only a temporary patch until the LinearIndices support is done
+    function Base.collect(inds::CartesianIndices)
+        dest = Array{eltype(inds)}(undef, size(inds))
+        @inbounds for (i, idx) in enumerate(inds)
+            dest[i] = idx
+        end
+        dest
+    end
```
I'm a bit worried about this:

```julia
julia> @btime collect(CartesianIndices((64, 64)));
  5.954 μs (2 allocations: 64.08 KiB) # this
  4.044 μs (2 allocations: 64.08 KiB) # master
```

We should fix this before this gets merged.

johnnychen94 committed:
add UnitCartesianIndices alias

johnnychen94 committed:
fix LinearIndices test bugs

johnnychen94 committed:
fix a dead dispatching branch and add more comments

johnnychen94 committed:
clean up and add news

johnnychen94 commented:
@timholy I think this PR is ready for a hopefully final review

> One thing we can probably do to ease the transition is define
> ```julia
> const UnitCartesianIndices{N,R<:NTuple{N,AbstractUnitRange{Int}}} = CartesianIndices{N,R}
> ```
> and encourage people to use it when they want to enforce a step size of 1.

I finally choose to not add this because I didn't find a use case in this PR. Most methods are dispatched at the range type level and not the CartesianIndices type level.

Other noteworthy changes since the last review:

* I first implemented a [quite complicated overflow_check funcion](https://github.com/JuliaLang/julia/pull/37829/commits/4b47f634241b22e01bd8bbd3084d272033cf56dc#diff-89e97f67f8058d32eac8d917db5bbdbfR424-R433) and then reverted it to keep consistent with current range behavior. (CRef #37876)
* `step(::CartesianIndices)` is added

timholy commented on NEWS.md:
```diff
@@ -117,6 +117,9 @@ Standard library changes
 * The `Pkg.Artifacts` module has been imported as a separate standard library.  It is still available as
   `Pkg.Artifacts`, however starting from Julia v1.6+, packages may import simply `Artifacts` without importing
   all of `Pkg` alongside. ([#37320])
+* `CartesianIndices` can now be constructed from three `CartesianIndex`es `I`, `S`, `J` with `I:S:J`.
```
Before you say how to construct them, maybe it would be best to say that `CartesianIndices` now supports step different from 1?

timholy commented on base/multidimensional.jl:
```diff
@@ -440,44 +481,76 @@ module IteratorsMD
+
+    # This is currently needed because converting to LinearIndices is only available when steps are
+    # all 1
+    # NOTE: this is only a temporary patch and could be possiblely removed when StepRange support to
```
```suggestion
    # NOTE: this is only a temporary patch and could be possibly removed when StepRange support to
```

timholy commented on base/multidimensional.jl:
```diff
@@ -440,44 +481,76 @@ module IteratorsMD
+    end
+    function Base.LinearIndices(inds::CartesianIndices)
+        indices = inds.indices
+        if all(x->x==1, step.(indices))
```
```suggestion
        if all(x->step(x)==1, indices)
```
is friendlier to the compiler, and it's a little weird to mix functional with the broadcasting. Your version is intermediate between mine and full-broadcasting, `all(step.(indices) .== 1)`.

timholy commented on base/multidimensional.jl:
```diff
@@ -440,44 +481,76 @@ module IteratorsMD
+    function Base.LinearIndices(inds::CartesianIndices)
+        indices = inds.indices
+        if all(x->x==1, step.(indices))
+            indices = map(rng->first(rng):last(rng), indices)
```
Could use `_convert2ind` here?

johnnychen94 replied:
Hmmm, this one explicitly converts a StepRange with step 1 to a UnitRange.

timholy commented on base/multidimensional.jl:
```diff
@@ -440,44 +481,76 @@ module IteratorsMD
+    # all 1
+    # NOTE: this is only a temporary patch and could be possiblely removed when StepRange support to
+    # LinearIndices is done
+    function Base.collect(inds::CartesianIndices)
```
I'm not sure we need this...won't the default implementation give exactly the same result?

johnnychen94 replied:
CartesianIndices with StepRange can't be converted to LinearIndices for now in the generic version: https://github.com/JuliaLang/julia/blob/3559b720d12c4b5b870dc881272c30dbec3653fe/base/abstractarray.jl#L909-L911

This version is modified from the logic there. I've thought about specializing on `StepRange` but it seems blocks other potential range types.

timholy reviewed (commented):
Looks good...after a couple more tweaks I think we can merge this.

johnnychen94 committed:
apply suggestions

johnnychen94 committed:
add the fast path for AbstractUnitRange

timholy commented:
Let's find out if generalizing this type causes package problems: @nanosoldier `runtests(ALL, vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/37829#issuecomment-706126360) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/4086959a886ed94fbfd32417489e909b652a9952/pkgeval/by_hash/3386b42_vs_3559b72/report.md). cc @maleadt

timholy commented:
Interesting failure in https://github.com/JuliaCI/NanosoldierReports/blob/4086959a886ed94fbfd32417489e909b652a9952/pkgeval/by_hash/3386b42_vs_3559b72/logs/Images/1.6.0-DEV-84648ad804.log

johnnychen94 commented:
I think that's due to https://github.com/JuliaArrays/TiledIteration.jl/blob/master/src/TiledIteration.jl#L102 and is fixable.

timholy commented:
Good. That seems like the only real one I've seen. I think we can merge this.

timholy commented:
Thank you @johnnychen94! :fireworks:

johnnychen94 commented:
My very first news-worth PR :tada:
